Support provider fully-qualified class name in Restricted Security mode

taoliult opened the pull request:
This PR is for updating the codes to only support the fully-qualified class name when setting the providers list in restricted security profiles. For example, the NSS FIPS140-2 provider list will be:

```
RestrictedSecurity.NSS.140-2.jce.provider.1 = sun.security.pkcs11.SunPKCS11 ${java.home}/conf/security/nss.fips.cfg
RestrictedSecurity.NSS.140-2.jce.provider.2 = sun.security.provider.Sun [ \
    {CertificateFactory, X.509, ImplementedIn=Software}, \
    {CertStore, Collection, ImplementedIn=Software}, \
    {CertStore, com.sun.security.IndexedCollection, ImplementedIn=Software}, \
    {Policy, JavaPolicy, *}, {Configuration, JavaLoginConfig, *}, \
    {CertPathBuilder, PKIX, ValidationAlgorithm=RFC5280:ImplementedIn=Software}, \
    {CertPathValidator, PKIX, ValidationAlgorithm=RFC5280:ImplementedIn=Software}, \
    {KeyStore, PKCS12, *}]
RestrictedSecurity.NSS.140-2.jce.provider.3 = sun.security.ec.SunEC [{KeyFactory, EC, ImplementedIn=Software: \
    SupportedKeyClasses=java.security.interfaces.ECPublicKey|java.security.interfaces.ECPrivateKey: \
    KeySize=256}, {AlgorithmParameters, EC, *}]
RestrictedSecurity.NSS.140-2.jce.provider.4 = sun.security.ssl.SunJSSE
```

keithc-ca commented on closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java:
```diff
@@ -857,6 +863,63 @@ private void initConstraints() {
+            }
+
+            try {
+                // Locate and parse providerattributes section.
+                pAttributes = pAttributes.replaceAll("\\s", "");
+                int startIndex = pAttributes.indexOf("<providerattributes>");
+                int endIndex = pAttributes.indexOf("</providerattributes>");
```
I suggest that xml is not needed, nor is it very user-friendly in a properties file (i.e. `java.security`). The example in the description could be expressed more simply and clearly as:
```
RestrictedSecurity.NSS.140-2.attributes = \
    SunJGSS=sun.security.jgss.SunProvider, \
    SunSASL=com.sun.security.sasl.Provider
```
(possibly using a name other than `attributes`)

There are several references in comments to `providerattributes`; did you mean "provider attributes"?

taoliult replied:
I considered this simply format(you suggested) before, but after discussed with the team, we think of a more general format. Right now, this XML tag “<providerattributes>” is used for the mapping between provider name and its class name. But, in the future, we can extend this "<restrictedsecurityattributes>" tag, to add other XML tags inside this “<restrictedsecurityattributes>” tag.

keithc-ca replied:
Ok, but if you want to use xml, you should be using existing xml parsing classes.

keithc-ca replied:
I take that back: java.base can't assume, for example, that the java.xml module is present and so can't use that code to parse xml.

Perhaps there's a better solution that doesn't require dealing with the "simple" name of a provider. Most of the uses of `isProviderAllowed(String)` could be switched to use `isProviderAllowed(Class)`. Could the remaining cases use the latter, or at least the (fully-qualfied) class name? If so, we don't need to discuss the format of the `attributes` property value.

taoliult replied:
@keithc-ca 

If we only use provider class name to check allow or not-allow. That means, in java.security file, we have to use provider class name to configure the FIPS provider list. For example, when we want to add the SunSASL in the list, it will be like below:

`RestrictedSecurity.NSS.140-2.jce.provider.5=com.sun.security.sasl.Provider`

And then, in the `ProviderConfig` class, when loading this non-java.base provider, it will use `ServiceLoader` to load it, and due to “com.sun.security.sasl.Provider” is allowed (it can pass the `isProviderAllowed(Class)` check), so this provider class can be loaded. 

But after that, in `ProviderLoader.load()` method, it will use the provider name “SunSASL” to find the provider class object and return it [1]. But due to we did not use provider name in the java.security FIPS provider list, so the provider “SunSASL” can not be found and return (even this provider class is already loaded).

[1] - https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/openj9/src/java.base/share/classes/sun/security/jca/ProviderConfig.java#L375

And also, in this `ProviderLoader` class, it do had another method called `legacyLoad()`. It is designed to load the provider by using the provider class name. But when I tried this method for "SunSASL", I get the below error:

> ProviderList: provider configuration: [com.sun.security.sasl.Provider]
> ProviderConfig: Loading legacy provider: com.sun.security.sasl.Provider
> ProviderConfig: Error loading legacy provider com.sun.security.sasl.Provider
> java.security.PrivilegedActionException: java.lang.IllegalAccessException: Class sun/security/jca/ProviderConfig$ProviderLoader$1(java.base) can not access class com/sun/security/sasl/Provider(java.security.sasl) because module java.security.sasl does not export package com/sun/security/sasl to module java.base

So, if the provider module did not export package to java.base module, then this `legacyLoad()` won't work for that provider class.

taoliult replied:
@keithc-ca Any update or advice on this PR?

keithc-ca replied:
Please update the description of this pull request: the discussion of unqualified names and XML are no longer relevant.

taoliult replied:
Description updated.

keithc-ca reviewed (changes requested):
Please correct the typo in the commit message: "a attribute" -> "an attribute".

taoliult commented on closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java:
```diff
@@ -593,9 +593,11 @@ private static final class RestrictedSecurityProperties {
         String jdkSecureRandomProvider;
         String jdkSecureRandomAlgorithm;
 
+        private final Map<String, String> providerClassNameMap;
```
This Map. the provider class name is the key, and the provider name defined in its constructor is the value. If we call it "providerSimpleNamesMap", is that more appropriate?

taoliult commented:
@keithc-ca I replied the above questions, and any update or advice on this PR?

taoliult commented:
@keithc-ca 

I updated the codes, to avoid the legacyLoad() method from the ProviderConfig class. For now, for the providers which provider name and its class name are same, we still use the provider name as the provider list setting. For the SunSASL, since we already release this to the client by using the provider name for the setting, so I did not remove this provider mapping. For the other providers, which the provider name and its class name are different, we can only use the provider class as the provider list setting. For example, the SunJGSS and XMLDSig, we can set them as below:

```
RestrictedSecurity.NSS.140-2.jce.provider.1 = sun.security.jgss.SunProvider
RestrictedSecurity.NSS.140-2.jce.provider.2 = org.jcp.xml.dsig.internal.dom.XMLDSigRI
```

Please help to review and advise.

@jasonkatonica FYI.

keithc-ca commented on closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java:
```diff
@@ -941,6 +940,11 @@ boolean isRestrictedServiceAllowed(Service service) {
 
             Constraint[] constraints = providerConstraints.get(providerName);
 
+            if (constraints == null) {
+                // Continue checking if the class name is the key.
+                constraints = providerConstraints.get(providerClass);
+            }
```
Suggest this should check `providerClassName` first.

taoliult replied:
@keithc-ca check providerClassName first, you mean like below?

```
            Constraint[] constraints = providerConstraints.get(providerClassName);

            if (constraints == null) {
                // Continue checking if the provider name is the key.
                constraints = providerConstraints.get(providerName);
            }
```

keithc-ca replied:
Yes.

keithc-ca commented on closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java:
```diff
@@ -1070,12 +1084,12 @@ boolean isRestrictedProviderAllowed(String providerName) {
          */
-        private static String getProvidersSimpleName(String providerName) {
+        private static String getProvidersNameWithoutPackage(String providerName) {
             if (providerName.equals("com.sun.security.sasl.Provider")) {
```
I don't think we need or want to make SunSASL a special case any more.

taoliult replied:
@keithc-ca We already released a version which suggested the client to set using SunSASL in their java.security provider list. So, that is why I did not remove this SunSASL provider mapping.

keithc-ca replied:
With the change to `ProviderConfig.load()`, it should it not recognize both `SunSASL` and `com.sun.security.sasl.Provider`? I think so, which eliminates the need for a special case here.

taoliult replied:
@keithc-ca 

If we set `RestrictedSecurity.NSS.140-2.jce.provider.N=com.sun.security.sasl.Provider,` this works. Because in the ServiceLoader, before the provider initialized, it will use the package + class name to check, and due to we use the package + class name in the java.security provider list, the check will pass.

But if we set `RestrictedSecurity.NSS.140-2.jce.provider.N=SunSASL`, this won't work. Because in the ServiceLoader, before the provider initialized, it will use the class name to check, and due to we use the provider name in the java.security provider list, and for SunSASL, its provider name is different with its class name(after removing the package), so it won't pass the check in ServiceLoader. 

For the non-java.base providers, the check happens in the ServiceLoader using its package + class name. So, if the provider name is different with its class name(after removing the package). We have to set its package + class name in the java.security RestrictedSecurity provider list.

keithc-ca replied:
In retrospect, it was a mistake to try to support `RestrictedSecurity.NSS.140-2.jce.provider.N=SunSASL`.

taoliult commented:
@keithc-ca Codes updated according to code review, others replied the questions. Please help to review and advise.

keithc-ca commented on src/java.base/share/classes/sun/security/jca/ProviderConfig.java:
```diff
@@ -190,6 +185,14 @@ Provider getProvider() {
             if (!shouldLoad()) {
                 return null;
             }
+            // There is RestrictedSecurity check in the ServiceLoader before the
+            // provider is initialized. So the check has already occurred for the
+            // provider where provider != null.
```
I think a better place for this comment is before line 176.

taoliult commented:
@keithc-ca Codes updated according to code review, others replied the questions. Please help to review and advise.

taoliult commented:
@keithc-ca Codes updated according to today's meeting. For NSS FIPS 140-2, both provider name and fully-qualified class name are accepted. For OpenJCEPlus FIPS 140-3, only fully-qualified class name is accepted. And also, removed the mapping for SunSASL.

And, because of the ProviderConfig updates need to call the method "RestrictedSecurity.isFIPS1402()" from RestrictedSecurity.java, so this PR depends on the PR https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/793

@jasonkatonica FYI.

keithc-ca commented on src/java.base/share/classes/sun/security/jca/ProviderConfig.java:
```diff
@@ -372,8 +375,14 @@ public Provider load(String pn) {
                     }
-                    if (pName.equals(pn)) {
-                        return p;
+                    if (RestrictedSecurity.isFIPSEnabled() && !RestrictedSecurity.isFIPS1402()) {
+                        if (p.getClass().getName().equals(pn)) {
+                            return p;
+                        }
+                    } else {
+                        if (pName.equals(pn) || p.getClass().getName().equals(pn)) {
+                            return p;
+                        }
                     }
```
Suggest checking the class name first:
```java
                    if (p.getClass().getName().equals(pn)) {
                        return p;
                    } else if (pName.equals(pn)) {
                        if (!RestrictedSecurity.isFIPSEnabled() || RestrictedSecurity.isFIPS1402()) {
                            return p;
                        }
                    }
```

keithc-ca commented on closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java:
```diff
@@ -932,7 +931,9 @@ private void initConstraints() {
-            String providerName = service.getProvider().getName();
+            Provider provider = service.getProvider();
+            String providerName = provider.getName();
+            String providerClassName = provider.getClass().getName();
```
Suggest updating the local in `isProviderAllowed(Class<?>)` from `providerName` to `providerClassName` to be consistent with this method.

keithc-ca commented on closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java:
```diff
@@ -1043,10 +1049,20 @@ boolean isRestrictedProviderAllowed(String providerName) {
+            }
+
+            // Provider name without package name.
+            providerName = getProvidersNameWithoutPackage(providerName);
```
I don't think we want this any more. Only the name as specified in the properties file (less any argument) should be used. The name of the field, `providersSimpleName`, should be changed: perhaps `providers` and `providersSimpleName`
should be `providerWithArg` and `providerWithoutArg`, respectively.

taoliult replied:
@keithc-ca 

I remember in our last meeting, for the NSS FIPS 140-2 provider list, we still use the provider name (the name from its constructor) in the provider list. Only the OpenJCEPlus FIPS 140-3 and any new providers which need to be added into the NSS FIPS 140-2 provider list, for example the SunSASL, will use the provider fully qualified name.

```
RestrictedSecurity.NSS.140-2.jce.provider.1 = SunPKCS11 ${java.home}/conf/security/nss.fips.cfg
RestrictedSecurity.NSS.140-2.jce.provider.2 = SUN [{CertificateFactory, X.509, ImplementedIn=Software}, \
    {CertStore, Collection, ImplementedIn=Software}, \
    {CertStore, com.sun.security.IndexedCollection, ImplementedIn=Software}, \
    {Policy, JavaPolicy, *}, {Configuration, JavaLoginConfig, *}, \
    {CertPathBuilder, PKIX, ValidationAlgorithm=RFC5280:ImplementedIn=Software}, \
    {CertPathValidator, PKIX, ValidationAlgorithm=RFC5280:ImplementedIn=Software}, \
    {KeyStore, PKCS12, *}]
RestrictedSecurity.NSS.140-2.jce.provider.3 = SunEC [{KeyFactory, EC, ImplementedIn=Software: \
    SupportedKeyClasses=java.security.interfaces.ECPublicKey|java.security.interfaces.ECPrivateKey: \
    KeySize=256}, {AlgorithmParameters, EC, *}]
RestrictedSecurity.NSS.140-2.jce.provider.4 = SunJSSE
```

So, for example, the SunEC, when the ServiceLoader check the sun.security.ec.SunEC, we still need to get its provider name SunEC from removing the package. So, that is why I did not remove this getProvidersNameWithoutPackage() method.

@jasonkatonica FYI.

keithc-ca replied:
My understanding is that we agreed that only fully-qualified class names would be allowed when specifying a restricted security profile (otherwise the restrictions can be bypassed). Yes, this may mean some migration work for users, but the longer we wait to make that change, the greater number of users will be impacted.

taoliult commented:
@keithc-ca 

Codes updated according to code review, others replied the questions. Please help to review and advise.

And I also added "isFIPS1402()" method from Allow extension of RestrictedSecurity profile PR https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/793. So, this PR won’t have the dependency on Allow extension of RestrictedSecurity profile PR.

taoliult commented:
@keithc-ca As we talked today, I removed the checks using the provider name, only use the fully qualified class name when checking the providers and services, for all the restricted security profiles. The changes I made as below:

1. Removed the method isFIPS1402() since we don’t need it.
2. Changed the variable name providersSimpleName to providersFullyQualifiedClassName.
3. Removed the codes about “remove argument -NSS-FIPS”, since we only use the fully qualified class name.
4. Removed the method getProvidersNameWithoutPackage() since we don’t need to remove the package.

Please help to review and advise.

@jasonkatonica FYI.

keithc-ca commented on closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java:
```diff
@@ -758,15 +759,7 @@ private void initProviders() {
                 }
                 providerName = providerName.trim();
 
-                // Remove argument, e.g. -NSS-FIPS, if present.
-                pos = providerName.indexOf('-');
-                if (pos >= 0) {
-                    providerName = providerName.substring(0, pos);
-                }
```
Why don't we need to keep this?

taoliult replied:
Revert them, done.

keithc-ca replied:
What about providers that must specify an argument, like in
```
RestrictedSecurity.NSS.140-2.securerandom.provider = SunPKCS11-NSS-FIPS
```
The reference to `SunPKCS11` should probably use the fully-qualified name, but `-NSS-FIPS` still needs to be removed to derive the class name.

taoliult replied:
I don't think we can set as below:

`RestrictedSecurity.NSS.140-2.securerandom.provider = SunPKCS11-NSS-FIPS`

The provider name SunPKCS11-NSS-FIPS comes from two parts, provider name defined in its constructor + name defined in its configure file, the "${java.home}/conf/security/nss.fips.cfg" file. The SunPKCS11 constructor codes below:

```
    // Used by Secmod
    SunPKCS11(Config c) {
        super("SunPKCS11-" + c.getName(), PROVIDER_VER, c.getDescription());
        this.config = c;
```

When this provider is initialized with the configure file, for example the "nss.fips.cfg" file, the name will be SunPKCS11-NSS-FIPS. Without configure file, the name will be SunPKCS11.

The argument "-NSS-FIPS" is not a part of the name in its constructor. So I don't think we can specify the argument (without using the configure file) in the java.security provider list.

keithc-ca replied:
My point is that the entry added to `providersFullyQualifiedClassName` is expected to be a fully-qualified class name, which would be `sun.security.pkcs11.SunPKCS11`, not `sun.security.pkcs11.SunPKCS11-NSS-FIPS`.

taoliult replied:
Yes, that is correct. For the SunPKCS11 in NSS FIPS140-2 profile, it would be:

`RestrictedSecurity.NSS.140-2.jce.provider.1 = sun.security.pkcs11.SunPKCS11 ${java.home}/conf/security/nss.fips.cfg`

keithc-ca replied:
Perhaps you misunderstand my concern. An existing properties file it might say:
```
RestrictedSecurity.NSS.140-2.jce.provider.1 = SunPKCS11-NSS-FIPS ...
```
I expected that this change will require that the class name will be fully-qualified, but the argument would still be present:
```
RestrictedSecurity.NSS.140-2.jce.provider.1 = sun.security.pkcs11.SunPKCS11-NSS-FIPS ...
```
Are you saying that must the argument be omitted?
```
RestrictedSecurity.NSS.140-2.jce.provider.1 = sun.security.pkcs11.SunPKCS11 ...
```
If so, it would seem there's no way to disallow other configurations of a given provider.

It's also not clear to me how the existing `java.properties` file could work properly for `SunPKCS11` because it does not specify the argument and so `RestrictedSecurity.isProviderAllowed("SunPKCS11-NSS-FIPS")` would return `false`.

keithc-ca commented on closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java:
```diff
@@ -1037,18 +1026,9 @@ boolean isRestrictedProviderAllowed(String providerName) {
-            // Check if the provider is in restricted security provider list.
-            // If not, the provider won't be registered.
-            if (providersSimpleName.contains(providerName)) {
+            // Check if the provider fully qualified cLass name is in restricted
```
Missing hyphen in "fully-qualified".

taoliult replied:
Done.

taoliult commented:
@keithc-ca Codes updated according to review, please help to review and advise.

keithc-ca commented on closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java:
```diff
@@ -932,7 +931,8 @@ private void initConstraints() {
+            Provider provider = service.getProvider();
+            String providerName = provider.getClass().getName();
 
             // Provider with argument, remove argument.
             // e.g. SunPKCS11-NSS-FIPS, remove argument -NSS-FIPS.
```
This no longer makes sense (`providerName` is a fully-qualified class name).

taoliult replied:
Yes, because we get the provider class name from the `Service`, so no need to remove argument. I removed these lines. 

And I also removed the "Remove argument" codes from the `initProviders()` method, because we only set the  fully-qualified class name in the java.security profile, so no need to remove the argument.

keithc-ca commented on closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java:
```diff
@@ -1043,12 +1043,9 @@ boolean isRestrictedProviderAllowed(String providerName) {
                 providerName = providerName.substring(0, pos);
```
The parameter of this method should be renamed to `providerClassName` to more accurately reflect its meaning.

The removal of the argument does not belong here, but should instead be in `isProviderAllowed(String)` (the only place it might not already be a fully-qualified class name).

taoliult replied:
Yes, and I also renamed to `providerClassName` in method `isRestrictedServiceAllowed(Service service)`.

keithc-ca commented:
> Please update the description of this pull request: the discussion of unqualified names and XML are no longer relevant.

See https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/758#discussion_r1643165104.

keithc-ca commented:
Please squash.

taoliult commented:
@keithc-ca Squashed, please help to review and advise.

keithc-ca commented:
Jenkins test sanity alinux jdknext

taoliult commented:
> @taoliult Please confirm that this change need not be back-ported to https://github.com/ibmruntimes/openj9-openjdk-jdk8.

@keithc-ca We need to back-ported it to Semeru OpenJDK8, working on it and will create the PR shortly.
